queue_consumer.go

```diff
@@ -121,8 +128,20 @@ func (mf *Consumer) receiveMessages(ctx context.Context, wg *sync.WaitGroup, ch
 	}
 }
 
-// Run starts the Consumer, gracefully stopping it when the given context is cancelled.
+// Run starts the Consumer, stopping it when the given context is cancelled.
+// To shut down without canceling the Context, and allow in-flight messages to drain,
+// call ShutDown.
+//
+// If ShutDown has been called before Run, the returned error is ErrConsumerAlreadyShutDown.
+// If the context is canceled, the returned error is the context's error.
+// If in-flight messages drain to completion, the returned error is nil.
 func (mf *Consumer) Run(ctx context.Context) error {
+	select {
+	case <-mf.shutDown:
+		return ErrConsumerAlreadyShutDown
```

Comment:
I would consider a panic here, since it's significantly incorrect behaviour and probably not recoverable. A stack trace is more likely to be useful than a log message.

Reply:
I think we can simplify this by stopping when the channel is closed.  If we get a closed channel to begin with, we "stop" without doing any work.  This way the code paths remain the same and there are no special cases the caller has to deal with.

We usually wire our services up to listen to SIGINT and SIGTERM.  It's possible for one of those to arrive before the service has finished setting up.  We _could_ require the caller to only wire up signal handling at the end of the setup, but that seems unnecessary.

Reply:
I don't think returning without an error is the right answer. I would choose between returning an error or panicking, so the caller gets some sort of diagnostic information if they try to run an already shutdown consumer. If you don't want to panic, then I'd leave this as is.

Reply:
Ok, I would prefer to return an error, given the scenario I mentioned above.

Reply:
I see your point. The caller at least has the ability to know why `Run` returned.